<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- NOTE: joint must match the name of the module -->
  <xacro:macro name="actuator" params="name child type">
  <xacro:property name="model_name" value="${type}"/>
    <link name="${name}/INPUT_INTERFACE">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <xacro:if value="${model_name.startswith('X5')}">
            <mesh filename="package://hebi_description/meshes/x5.stl" scale="0.001 0.001 0.001"/>
          </xacro:if>
          <xacro:if value="${model_name.startswith('X8')}">
            <mesh filename="package://hebi_description/meshes/x8.stl" scale="0.001 0.001 0.001"/>
          </xacro:if>
        </geometry>
        <xacro:red_material/>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <xacro:if value="${model_name.startswith('X5')}">
            <mesh filename="package://hebi_description/meshes/x5.stl" scale="0.001 0.001 0.001"/>
          </xacro:if>
          <xacro:if value="${model_name.startswith('X8')}">
            <mesh filename="package://hebi_description/meshes/x8.stl" scale="0.001 0.001 0.001"/>
          </xacro:if>
        </geometry>
      </collision>
      <inertial>
        <xacro:if value="${model_name.startswith('X5')}">
          <origin xyz="-0.0142 -0.0031 0.0165" rpy="0 0 0" />
          <xacro:if value="${model_name == 'X5_1'}">
            <mass value="0.315"/>
          </xacro:if>
          <xacro:if value="${model_name == 'X5_4'}">
            <mass value="0.335"/>
          </xacro:if>
          <xacro:if value="${model_name == 'X5_9'}">
            <mass value="0.360"/>
          </xacro:if>
          <inertia ixx="0.00015" iyy="0.000255" izz="0.000350" ixy="0.0000341" ixz="0.0000118" iyz="0.00000229"/>
        </xacro:if>
        <xacro:if value="${model_name.startswith('X8')}">
          <origin xyz="-0.0145 -0.0031 0.0242" rpy="0 0 0" />
          <xacro:if value="${model_name == 'X8_3'}">
            <mass value="0.460"/>
          </xacro:if>
          <xacro:if value="${model_name == 'X8_9'}">
            <mass value="0.480"/>
          </xacro:if>
          <xacro:if value="${model_name == 'X8_16'}">
            <mass value="0.500"/>
          </xacro:if>
          <inertia ixx="0.000246" iyy="0.000380" izz="0.000463" ixy="0.0000444" ixz="0.0000266" iyz="0.00000422"/>
        </xacro:if>
      </inertial>
    </link>
    <joint name="${name}/${model_name}" type="continuous">
      <parent link="${name}/INPUT_INTERFACE"/>
      <child link="${child}/INPUT_INTERFACE"/>
      <xacro:if value="${model_name.startswith('X5')}">
        <origin xyz="0 0 0.031" rpy="0 0 0"/>
      </xacro:if>
      <xacro:if value="${model_name.startswith('X8')}">
        <origin xyz="0 0 0.045" rpy="0 0 0"/>
      </xacro:if>
      <axis xyz="0 0 1"/>
      <xacro:if value="${model_name == 'X5_1'}">
        <limit effort="2.5" velocity="90"/>
        <dynamics damping="0.1" friction="0.1"/>
      </xacro:if>
      <xacro:if value="${model_name == 'X5_4'}">
        <limit effort="7.0" velocity="32"/>
        <dynamics damping="0.1" friction="0.1"/>
      </xacro:if>
      <xacro:if value="${model_name == 'X5_9'}">
        <limit effort="13.0" velocity="14"/>
        <dynamics damping="0.1" friction="0.1"/>
      </xacro:if>
      <xacro:if value="${model_name == 'X8_3'}">
        <limit effort="7" velocity="84"/>
        <dynamics damping="0.5" friction="0.5"/>
      </xacro:if>
      <xacro:if value="${model_name == 'X8_9'}">
        <limit effort="20" velocity="30"/>
        <dynamics damping="0.5" friction="0.5"/>
      </xacro:if>
      <xacro:if value="${model_name == 'X8_16'}">
        <limit effort="38" velocity="15"/>
        <dynamics damping="0.5" friction="0.5"/>
      </xacro:if>
    </joint>
  </xacro:macro>
</robot>

